datafusion/src/optimizer/filter_push_down.rs

```diff
@@ -408,10 +412,20 @@ fn optimize(plan: &LogicalPlan, mut state: State) -> Result<LogicalPlan> {
             //
             // Join clauses with `Using` constraints also take advantage of this logic to make sure
             // predicates reference the shared join columns are pushed to both sides.
+            let mut null_unsafe_predicates = vec![];
             let join_side_filters = state
                 .filters
                 .iter()
                 .filter_map(|(predicate, columns)| {
+                    if join_type == &JoinType::Left
+                        || join_type == &JoinType::Right
+                        || join_type == &JoinType::Full
+                    {
+                        if let Expr::IsNull(..) | Expr::IsNotNull(..) = predicate {
```

Comment:
Should we check that the expression to *contains* `is null` and `is not null`?
Silly example:  `1=1 AND x is null` I think should have the same effect as `x is null` for denying the optimization.

Reply:
Nice caught! I'll add a function to check if an `Expr` contains `IsNotNuLL` or `IsNull`

Reply:
I haven't had a chance to study this PR yeet in depth, but I wanted to point out a few things:

I don't think we can push *ANY* filers from the `ON` clause down the non preserved side(s) of an outer join -- in other words, this is not a problem with `IS NULL` and `IS NOT NULL`.

For example

```sql
alamb=# select * from A;
 a 
---
 1
 2
(2 rows)

alamb=# select * from B;
 b 
---
 1
 2
(2 rows)
```

Note if you pushed the predicate `a>5` down, there would be no rows, but the output is:
```sql
alamb=# select * from A LEFT JOIN B ON (a=b and a>5);
 a | b 
---+---
 1 |  
 2 |  
```

Note the semantics of predicates in the `WHERE` clause are different (they are always pushed down)

```
alamb=# select * from A LEFT JOIN B ON (a=b) WHERE  a>5;
 a | b 
---+---
(0 rows)
```


However, note you *CAN* push `b>5` down and get the correct output (because b is not preserved):

```sql
alamb=# select * from A LEFT JOIN B ON (a=b and b>5);
 a | b 
---+---
 1 |  
 2 |  
(2 rows)
```

The criteria is more like "refers to a column on the non preserved side"

Reply:
> I don't think we can push _ANY_ filers from the `ON` clause down the non preserved side(s) of an outer join -- in other words, this is not a problem with `IS NULL` and `IS NOT NULL`.

Yes, I agree. Luckily, we don't implement the case. https://github.com/apache/arrow-datafusion/blob/0facd4d483e8c289ee4e3a89487d0cd1ede1a110/datafusion/src/sql/planner.rs#L564

However, I think there is a problem with `IS NULL` and `IS NOT NULL` in `WHERE` clause, we shouldn't push them down.